validators/cgroup_validator_linux.go

```diff
@@ -142,15 +142,19 @@ func (c *CgroupsValidator) getCgroupV2Subsystems() ([]string, error) {
 	// Some controllers are implicitly enabled by the kernel.
 	// Those controllers do not appear in /sys/fs/cgroup/cgroup.controllers.
 	// https://github.com/torvalds/linux/blob/v5.3/kernel/cgroup/cgroup.c#L433-L434
-	// We assume these are always available, as it is hard to detect availability.
-	// So, we hardcode the following as "pseudo" controllers.
+	// For freezer which is implemented in kernel 5.2, we can check the existence of `cgroup.freeze`.
+	// For others, we assume these are always available, as it is hard to detect availability.
+	// We hardcode the following as initial controllers.
 	// - devices: implemented in kernel 4.15
-	// - freezer: implemented in kernel 5.2
-	pseudo := []string{"devices", "freezer"}
+	subsystems := []string{"devices"}
+	_, err := os.Stat("/sys/fs/cgroup/init.scope/cgroup.freeze")
```

Comment:
It might be a corner case, but we want to get a path of cgroup.freeze
in a way that is independent of the environment.
- "/sys/fs/cgroup" is the default mountpoint and modifiable.
- "init.scope" is specific for systemd.

Reply:
> * "/sys/fs/cgroup" is the default mountpoint and modifiable.

https://github.com/kubernetes/system-validators/blob/1d2f5889f07347b9f2d3ed337b3cfb08d45f4cf6/validators/cgroup_validator_linux.go#L48

We currently use this default here for other checks.

If not, we may have to check the `mount`  or `/proc/mounts`/`/proc/self/mounts` to get the cgroup2 mount path. 


> "init.scope" is specific for systemd.

Or we may use `system.slice/cgroup.freeze`?

Reply:
```
	_, err := os.Stat(filepath.Join(unifiedMountpoint, "system.slice", "/cgroup.freeze"))
```
Changed to this. 

https://github.com/kubernetes/system-validators/blob/4d61a4e8f440a37c88e281ae115e43e8893683a1/validators/cgroup_validator_linux.go#L150-L154

Reply:
Thank you for your suggestion.

> If not, we may have to check the `mount` or `/proc/mounts`/`/proc/self/mounts` to get the cgroup2 mount path.

Yes. I think we need to get the path from /proc/mounts.
@neolit123  Is this change ok? It may affect the behavior of existing check.

> Or we may use system.slice/cgroup.freeze?

system.slice is also dependent on systemd.
Creating a directory in cgroupfs for this check is the perfect solution.

Reply:
unfortunately i don't know a lot about the details here.

one point to make - kubeadm supports the openrc init system from alpine and not only systemd:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/util/initsystem/initsystem_unix.go#L46C14-L46C20

also, k8s kubelet currently only works with the systemd cgroups driver AFAIK.
but for runc it's not mandatory:
https://github.com/opencontainers/runc/blob/main/docs/cgroup-v2.md#systemd 

> system.slice is also dependent on systemd.

yes there are still users using the cgroupfs driver, which means that system.slice will not exist.

> Creating a directory in cgroupfs for this check is the perfect solution.

could you elaborate why we need to create a dir?

IIUC what we need to do to check if a controller foo is enabled is to do the following:
```
check if these paths exist:
  - for cgroupfs /sys/fs/cgroup/freezer
   - for systemd /sys/fs/cgroup/systemd/freezer.state
```
if one of them exists we can say freezer is enabled, i don't think we need to check freezer.state for values.

Reply:
Updated.

You can see https://github.com/kubernetes/system-validators/compare/4d61a4e8f440a37c88e281ae115e43e8893683a1..2ec937dbfb62d8fa673a16e73db0fa21eb042d3b. 

@KentaTada Do you have time to take another look?

Reply:
related
- https://github.com/k0sproject/k0s/issues/3069

Reply:
> could you elaborate why we need to create a dir?

/sys/fs/cgroup/freezer does not exist.
Like k0s, we also need to create a dummy cgroup:
https://github.com/k0sproject/k0s/pull/3077/files#diff-468315026edb0a5cee0863f23231ad03809dcf5aa33958e7df1fd03779aa2621R171

Reply:
> @KentaTada Do you have time to take another look?

I took another look.
Could you confirm it?